proposals/image-metadata.md

```diff
@@ -0,0 +1,81 @@
+
+The metadata is added to the image as a `devcontainer.metadata` label with a JSON string value representing the above array or single object.
+
+## Merge Logic
+
+To apply the metadata together with a user's devcontainer.json at runtime the following merge logic by property is used. The table also notes which properties are currently supported coming from the devcontainer.json and which from the feature metadata, this will change over time as we add more properties.
+
+| Property | Type/Format | Merge Logic | devcontainer.json | Feature Metadata |
+| -------- | ----------- | ----------- | :---------------: | :--------------: |
+| `id` | E.g., `ghcr.io/devcontainers/features/node:1` | Not merged. |   | x |
+| `init` | `boolean` | `true` if at least one is `true`, `false` otherwise. |   | x |
+| `privileged` | `boolean` | `true` if at least one is `true`, `false` otherwise. |   | x |
+| `capAdd` | `string[]` | Union of all `capAdd` arrays without duplicates. |   | x |
+| `securityOpt` | `string[]` | Union of all `securityOpt` arrays without duplicates. |   | x |
+| `entrypoint` | `string` | Collected list of all entrypoints. |   | x |
+| `mounts` | `(string \| { type, src, dst })[]` | Collected list of all mountpoints. Conflicts: Last source wins. | x | x |
+| `onCreateCommand` | `string \| string[]` | Collected list of all onCreateCommands. | x |   |
+| `updateContentCommand` | `string \| string[]` | Collected list of all updateContentCommands. | x |   |
+| `postCreateCommand` | `string \| string[]` | Collected list of all postCreateCommands. | x |   |
+| `postStartCommand` | `string \| string[]` | Collected list of all postStartCommands. | x |   |
+| `postAttachCommand` | `string \| string[]` | Collected list of all postAttachCommands. | x |   |
+| `customizations` | Object of tool-specific customizations. | Merging is left to the tools. | x | x |
+| `remoteUser` | `string` | Last value wins. | x |   |
+| `userEnvProbe` | `string` (enum) | Last value wins. | x |   |
+| `remoteEnv` | Object of strings. | Per variable, last value wins. | x |   |
+| `containerEnv` | Object of strings. | Per variable, last value wins. | x |   |
+| `overrideCommand` | `boolean` | Last value wins. | x |   |
+| `portsAttributes` | Map of ports to attributes. | Per port (not per port attribute), last value wins. | x |   |
+| `otherPortsAttributes` | Port attributes. | Last value wins (not per port attribute). | x |   |
+| `forwardPorts` | `(number \| string)[]` | Union of all ports without duplicates. Last one wins (when mapping changes). | x |   |
+| `shutdownAction` | `string` (enum) | Last value wins. | x |   |
+| `updateRemoteUserUID` | `boolean` | Last value wins. | x |   |
+| `hostRequirements` | `cpus`, `memory`, `storage` | Max value wins. | x |   |
+
+Variables in string values will be substituted at the time the value is applied.
+
+## Notes
```

Comment:
How do we communicate to the user (someone using Codespaces or Remote-Containers) that the image they are referencing in the `image` property is "special" and will be applying additional dev container configuration into their dev environment?

After this change is applied, images under `mcr.microsoft.com/devcontainers` will behave differently than before (eg: the `remoteUser` will be applied in cases where it previously didn't and the user was defaulted to root). 

Just something to think about 🤔

Reply:
I agree. We could show additional information in a hover on the "dockerfile" or "image" properties in the devcontainer.json.

Reply:
There's this really slick looking, non-intrusive "info" diagnostic that vscode provides (the three grey dots underneath the string)

<img width="688" alt="image" src="https://user-images.githubusercontent.com/23246594/190502391-12c03d05-cdf8-4e79-8378-ee8dc1726a14.png">

Would be cool to provide info here on the config that will be applied!

Reply:
this and some useful logging about applying the labels as config might be enough